File: 1963. Minimum Number of Swaps to Make the String Balanced/MinSwaps.py
import math


class Solution:
    
    # O(n) time,
    # O(n) space,
    # Approach: greedy, counting
    def minSwaps(self, s: str) -> int:
        right_op = len(s)-1
        while s[right_op] != '[':
            right_op -= 1
        
        s = list(s)
        swaps = 0
        diff = 0
        for i in range(len(s)):
            if s[i] == ']':
                diff -= 1
            else:
                diff += 1
                
            if diff < 0:
                s[i], s[right_op] = s[right_op], s[i]
                right_op -= 1
                while s[right_op] != '[':
                    right_op -= 1
                swaps += 1
                diff += 2
                
        return swaps
    
    
    # O(n) time,
    # O(1) space,
    # Approach: counting, greedy, 
    def minSwaps2(self, s: str) -> int:
        extra_closed = 0
        
        max_closed = 0
        for ch in s:
            if ch == ']':
                extra_closed += 1
            else:
                extra_closed -= 1   
            max_closed = max(max_closed, extra_closed)
            
        return math.ceil(max_closed/2)

File: 1963. Minimum Number of Swaps to Make the String Balanced/minSwaps.go
package MinSwapsToMakeBalancedString

func shiftRightToOpeningBracket(chars []rune, index int) int {
	for index >= 0 && byte(chars[index]) != '[' {
		index -= 1
	}

	return index
}

// O(n) time,
// O(n) space,
// Approach: counting, two pointers
func minSwaps(s string) int {
	chars := []rune(s)
	lastOpeningIndex := shiftRightToOpeningBracket(chars, len(s)-1)

	closings := 0
	swaps := 0
	for i := 0; i < len(s); i++ {
		curr := chars[i]
		if curr == ']' {
			closings += 1
		} else {
			closings -= 1
		}
		if closings > 0 {
			swaps += 1
			closings -= 2
			chars[i], chars[lastOpeningIndex] = chars[lastOpeningIndex], chars[i]
			lastOpeningIndex = shiftRightToOpeningBracket(chars, lastOpeningIndex)
		}
	}

	return swaps
}
